src/zulip/mod.rs

```diff
@@ -0,0 +1,191 @@
+            user_group_ids,
+            zulip_api,
+            dry_run,
+        })
+    }
+
+    /// Create or update a user group for the given team name and members
+    fn create_or_update_user_group(
+        &mut self,
+        user_group_name: &str,
+        member_zulip_ids: &[usize],
+    ) -> Result<(), Error> {
+        let id = self.user_group_id_from_name(&user_group_name);
+        let user_group_id = match id {
+            Some(id) => {
+                log::info!(
+                    "'{}' user group ({}) already exists on Zulip",
+                    user_group_name,
+                    id
+                );
+                id
+            }
+            None => {
+                log::info!(
+                    "no '{}' user group found on Zulip. Creating one...",
+                    user_group_name
+                );
+                self.create_user_group(
+                    &user_group_name,
+                    &format!("The {} team", user_group_name),
+                    member_zulip_ids,
+                )?;
+                return Ok(());
+            }
+        };
+
+        let existing_members = self.user_group_members_from_name(&user_group_name).unwrap();
+        log::info!(
+            "'{}' user group ({}) has members on Zulip {:?} and needs to have {:?}",
+            user_group_name,
+            user_group_id,
+            existing_members,
+            member_zulip_ids
+        );
+        let add_ids = member_zulip_ids
+            .iter()
+            .filter(|i| !existing_members.contains(i))
+            .copied()
+            .collect::<Vec<_>>();
+        let remove_ids = existing_members
+            .iter()
+            .filter(|i| !member_zulip_ids.contains(i))
+            .copied()
+            .collect::<Vec<_>>();
+
+        // We don't currently update the members field of the cached user group because it's
+        // not necessary, but for correctness sake we should consider doing so
+        self.zulip_api
+            .update_user_group_members(user_group_id, &add_ids, &remove_ids)
+    }
+
+    /// Get a user group id for the given user group name
+    fn user_group_id_from_name(&self, user_group_name: &str) -> Option<usize> {
+        self.user_group_ids.get(user_group_name).map(|u| u.id)
+    }
+
+    /// Create a user group with a certain name, description, and members
+    fn create_user_group(
+        &mut self,
+        user_group_name: &str,
+        description: &str,
+        member_ids: &[usize],
+    ) -> Result<usize, Error> {
+        self.zulip_api
+            .create_user_group(user_group_name, description, member_ids)?;
+
+        // Update the user group cache so it has the user group that was just created
+        let user_groups = self.zulip_api.get_user_groups()?;
+        let user_groups = user_groups
+            .into_iter()
+            .map(|ug| (ug.name.clone(), ug))
+            .collect();
+        self.user_group_ids = user_groups;
```

Comment:
Should we do this only when we're not in a dry run? Otherwise this would override the groups previously added to the cache by the code below.